apps/src/lib/client/tx.rs

```diff
@@ -779,33 +777,37 @@ async fn is_safe_voting_window(
 /// vote)
 async fn filter_delegations(
     client: &HttpClient,
-    mut delegation_addresses: Vec<Address>,
+    delegations: HashSet<Address>,
     proposal_id: u64,
     delegator_vote: &ProposalVote,
-) -> Vec<Address> {
-    let mut remove_indexes: Vec<usize> = vec![];
-
-    for (index, validator_address) in delegation_addresses.iter().enumerate() {
-        let vote_key = gov_storage::get_vote_proposal_key(
-            proposal_id,
-            validator_address.to_owned(),
-            validator_address.to_owned(),
-        );
-
-        if let Some(validator_vote) =
-            rpc::query_storage_value::<ProposalVote>(client, &vote_key).await
-        {
-            if &validator_vote == delegator_vote {
-                remove_indexes.push(index);
-            }
-        }
-    }
-
-    for index in remove_indexes {
-        delegation_addresses.swap_remove(index);
-    }
+) -> HashSet<Address> {
+    // Filter delegations by their validator's vote concurrently
+    let delegations = futures::future::join_all(
+        delegations
+            .into_iter()
+            // we cannot use `filter/filter_map` directly because we want to
+            // return a future
+            .map(|validator_address| async {
+                let vote_key = gov_storage::get_vote_proposal_key(
+                    proposal_id,
+                    validator_address.to_owned(),
+                    validator_address.to_owned(),
+                );
 
-    delegation_addresses
+                if let Some(validator_vote) =
+                    rpc::query_storage_value::<ProposalVote>(client, &vote_key)
+                        .await
+                {
+                    if &validator_vote == delegator_vote {
+                        return None;
+                    }
+                }
+                Some(validator_address)
+            }),
+    )
+    .await;
+    // Take out the `None`s
+    delegations.into_iter().flatten().collect()
```

Comment:
very nice!

Reply:
interesting pattern, to turn an iterator into a future. I'll try to use this whenever I can :D